Add support for custom HTLC TLVs

alecchendev opened the pull request:
Closes #1298. This PR adds support for sending and receiving custom HTLC TLVs.

Custom TLVs allow users to send extra application-specific data with a payment. These have the additional flexibility compared to `payment_metadata` that they don't have to reflect recipient generated data provided in an invoice, in which `payment_metadata` could (probably) be reused.

On the send side, a user can provide their serialized TLVs as a `Vec<u8>` to `RecipientOnionFields::with_custom_tlvs` which checks whether it's a properly serialized TLV stream with type numbers in the range reserved for custom values. Then they'll pass this into whichever send payment function they use. This will then be serialized in the onion payload and sent with the payment.

On the receive side, when deserializing the onion payload, we just add the bytes back into a vector, then pipe this TLV stream from the onion through to `Event::PaymentClaimable`, dropping non-matching TLVs between payment parts.

There were a couple of rough edges while making this all work nicely--see the commit messages for elaboration :)

alecchendev commented on lightning/src/offers/merkle.rs:
```diff
@@ -144,7 +145,7 @@ fn tagged_branch_hash_from_engine(
 /// well-formed TLV stream.
 #[derive(Clone)]
-pub(super) struct TlvStream<'a> {
+pub(crate) struct TlvStream<'a> {
```
I tried to reuse these `TlvStream/TlvRecord` structs since it seemed I was going to have to implement something similar. Should I move this out to some other file since or is it fine here?

alecchendev commented on lightning/src/ln/msgs.rs:
```diff
@@ -1934,20 +1937,39 @@ impl Writeable for OnionHopData {
 	}
 }
 
+fn sorted_tlv_stream(bytes: &[u8]) -> Result<Vec<u8>, io::Error> {
+	let mut tlv_stream = TlvStream::try_new(bytes).map_err(|err| {
+		match err {
+			DecodeError::Io(kind) => io::Error::new(kind, kind.to_string()),
+			_ => io::Error::new(io::ErrorKind::Other, err.to_string()),
+		}
+	})?.collect::<Vec<TlvRecord>>();
```
Since writes are usually `expect/unwrap`'d, if custom TLVs are invalid here it'll panic while serializing `OnionHopData`. The user shouldn't be able to provide an invalid TLV stream since `RecipientOnionFields::with_custom_tlvs` should prevent that, but should we allow this? Thinking maybe should I add a check somewhere in the payment flow to return an error before this, or maybe on failure, omit the custom tlvs and go through with the payment?

alecchendev replied:
Became unnecessary when changing custom tlvs from a `Vec<u8>` to `Vec<(u64, Vec<u8>)>`

alecchendev reviewed (commented):
A couple questions for reviews:

alecchendev commented on lightning/src/ln/msgs.rs:
```diff
@@ -1956,14 +1978,27 @@ impl Readable for OnionHopData {
 			(16, payment_metadata, option),
 			// See https://github.com/lightning/blips/blob/master/blip-0003.md
 			(5482373484, keysend_preimage, option)
+		}, |msg_type: u64, msg_reader: &mut FixedLengthReader<_>| -> Result<bool, DecodeError> {
+			if msg_type < 1 << 16 { return Ok(false) }
```
Currently in this PR, as long as the custom TLVs are within the custom type number range, we'll succeed deserializing and provide the bytes to the user, regardless of whether the user actually recognizes these type numbers. I did this as a simple way to start, but IIUC we should actually fail deserialization if it's unknown + even. So should I add some way for the user to provide a custom handler here? Or at the least let them provide recognized type numbers?

alecchendev replied:
Yea sorry was going to come back to this.

> accept everything at the deserialization stage and then require the user call some claim_payment_with_known_custom_tlvs thing if they're even?

As in the normal `claim_funds` would fail if the attempted payment had even custom TLVs on it? And then `claim_payment_with_known_custom_tlvs` would just be the same thing except that it wouldn't fail, and would only be used when the user successfully recognized all the even TLVs?

Also I hadn't gotten around to it yet but @wpaulino had mentioned offline adding another `channelmanager::FailureCode` variant for the `invalid_onion_payload` error so the user could `ChannelManager::fail_htlc_backwards_with_reason` if they failed to recognize an even TLV - was planning to do this as well

alecchendev replied:
Added functionality for everything mentioned here

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -452,7 +464,20 @@ impl RecipientOnionFields {
+	/// deserialized properly into TLVs, otherwise returns the error faced while decoding. This
+	/// will also reject TLVs with type numbers outside the range reserved for custom types, i.e.
+	/// <2^16.
+	pub fn with_custom_tlvs(mut self, custom_tlvs: Vec<u8>) -> Result<Self, DecodeError> {
```
Just checking, I'm thinking `(T, V)` as `(u64, Vec<u8>)`? I was initially thinking `V: Writeable` but then all the values would have to be the same type, and doing something like `Box<dyn Writeable>` seemed overcomplicated

alecchendev replied:
Okay yea this simplified things a ton, thanks

alecchendev commented on lightning/src/offers/merkle.rs:
```diff
@@ -167,16 +185,46 @@ impl<'a> TlvStream<'a> {
 		self.filter(|record| !SIGNATURE_TYPES.contains(&record.r#type))
 	}
+
+	pub(crate) fn intersection(self, rhs: Self) -> Vec<u8> {
```
I could be wrong but I doubt this will ever be much of a performance bottleneck (and if it does it'll be pretty easy to fix in the future) so I decided to remove it

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -344,19 +346,37 @@ struct ReceiveError {
+	///
+	/// The tuple data should include the type number and byte offset in the decrypted byte stream
+	/// where the failure occurred.
+	InvalidOnionPayload(Option<(u64, u16)>),
```
So the user won't really have the ability to know the byte offset, but the type number could be helpful for the sender to debug. In most places where we return this error code we don't specify any data at all - wasn't sure if it makes sense to include this `Option<(u64, u16)>` in there or not?

alecchendev commented:
Rebased due to conflicts.

alecchendev commented on lightning/src/ln/msgs.rs:
```diff
@@ -1954,14 +1956,19 @@ impl Writeable for OnionHopData {
+					Some((5482373484, preimage.encode()))
+				} else { None };
+				let mut custom_tlvs: Vec<&(u64, Vec<u8>)> = custom_tlvs.iter().flatten().chain(preimage.iter()).collect();
+				custom_tlvs.sort_by_key(|(typ, _)| *typ);
```
Yea the custom TLVs are forced to be sorted in our public API, but in the case where some custom TLVs include type numbers both above and below keysend's type, we have to find the right place to put it

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7248,12 +7311,14 @@ impl_writeable_tlv_based_enum!(PendingHTLCRouting,
 		(1, phantom_shared_secret, option),
 		(2, incoming_cltv_expiry, required),
 		(3, payment_metadata, option),
+		(5, custom_tlvs, option),
```
> we should note in the release notes

Added a note to `pending_changelog` (seems like that's where this might go?) - noted it for 0.0.116 but ofc may end up being in 117

alecchendev commented on lightning/src/util/ser_macros.rs:
```diff
@@ -192,18 +192,35 @@ macro_rules! _get_varint_length_prefixed_tlv_length {
 		BigSize(len as u64).write($stream)?;
 		$crate::encode_tlv_stream!($stream, { $(($type, $field, $fieldty)),* });
-	} }
+		for (typ, value) in $extra_tlvs.iter() {
```
Done! Made a new macro for the implementation of `encode_tlv_stream` so that this new variant isn't exposed in the public API

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -110,13 +110,15 @@ pub(super) enum PendingHTLCRouting {
 		payment_preimage: PaymentPreimage,
 		payment_metadata: Option<Vec<u8>>,
 		incoming_cltv_expiry: u32, // Used to track when we should expire pending HTLCs that go unclaimed
+		custom_tlvs: Option<Vec<(u64, Vec<u8>)>>,
```
optional nit: I know it’s documented elsewhere, but maybe document the tuple here for easy reference.

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -455,7 +458,40 @@ impl RecipientOnionFields {
+		for (typ, _) in custom_tlvs.iter() {
+			if *typ < 1 << 16 { return Err(()); }
+			match prev_type {
+				Some(prev) if prev >= *typ => return Err(()),
```
If there are unsorted duplicates however, won't they get caught since this enforces that the types are strictly increasing?

alecchendev replied:
Huh yea good idea

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -455,7 +458,40 @@ impl RecipientOnionFields {
+	/// reserved for custom types, i.e. >= 2^16, otherwise this method will return `Err(())`.
+	/// See [`Self::custom_tlvs`] for more info.
+	pub fn with_custom_tlvs(mut self, custom_tlvs: Vec<(u64, Vec<u8>)>) -> Result<Self, ()> {
+		let mut prev_type = None;
```
Added

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -431,10 +431,13 @@ pub struct RecipientOnionFields {
 	/// may not be supported as universally.
 	pub payment_metadata: Option<Vec<u8>>,
+	/// See [`Self::custom_tlvs`] for more info.
+	pub(super) custom_tlvs: Option<Vec<(u64, Vec<u8>)>>,
```
Done!

alecchendev replied:
Squashed these changes to let CI run

alecchendev commented:
Rebased due to build errors, addressed feedback (most of the fixups are just removing the `Option` all over the place), and also fixed a bug: On retries I had left an empty value for custom TLVs as a placeholder - fixed this with https://github.com/lightningdevkit/rust-lightning/pull/2308/commits/c76ed85ff14cd3ec09833fdbddd882d7ee0256f1 and added a test in https://github.com/lightningdevkit/rust-lightning/pull/2308/commits/c2c5474a28b3e8c90052c1fab40bd0d939f5f3fd

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4748,6 +4801,23 @@ where
+
+				if let Some(RecipientOnionFields { ref custom_tlvs, .. }) = payment.onion_fields {
+					if !custom_tlvs_known && custom_tlvs.iter().any(|(typ, _)| typ % 2 == 0) {
+						log_info!(self.logger, "Cannot accept payment with unknown even TLVs. Rejecting payment with payment hash {}",
```
The unknown even TLVs could be printed in the `log_info` by displaying them as a vec.

ariard replied:
Cool.

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4715,13 +4746,35 @@ where
+	/// # Note
+	///
+	/// You MUST check you've understood all even TLVs before using this to
+	/// claim, otherwise you may unintentionally agree to some protocol you do not understand.
```
I don’t know if it should be documented here, in the pending changelog about this new custom HTLC TLVs support or at the spec-level, though as a receiver it should be mandatory to have a receiver-issued secret as part of the custom TLV before to even accept to process an experimental `tlv_record` from an _unknown_ LN sender.

See the concern above deanonymization attack, and how such requirement would serve as a deterrent.

ariard replied:
Let’s take the “Cancellable payments” protocol as described here: https://github.com/BlockstreamResearch/scriptless-scripts/blob/master/md/multi-hop-locks.md#cancellable-payments

Let’s assume there is a `valid_until_chain_height` negotiated as part of the protocol after which the multi-shard payments should be all cancelled by the recipient back to the sender.

Lack of authentication by the mean of a receiver-issued secret allow anyone with the details of the cancellable payments to send a payment shard (e.g a routing hop), and therefore to probe the chain height as seen by the recipient Lightning node (a fingerprint vector that could be used to cross-link with the on-chain full-node).

The landscape could be different with PTLCs, though here it’s too early to say if/how mutli-shard payment could be carried through. Yes, PTLCs could come with an inherent access structure, though this is not certain as of today.

Overall, it’s a reasonable internet protocol design rule to authenticate your network input as early as you can. Recommending our custom HTLC TLVs to authenticate its custom protocol sounds a good rule of thumb.

For recent failure between Lightning API interface and application processing, see https://lists.linuxfoundation.org/pipermail/lightning-dev/2023-June/003983.html

ariard replied:
> can you describe an attack where payment_secret isn't sufficient and something has to be added to the custom TLVs?

Yes where you have multiple payers and one payee, and each have to pay their shares of the payment e.g a crowdfunding between named participants. So current `payment_secret` wouldn’t work and you would need a specific authentication flow.

Somehow that type of multi-payer single-payee for crowdfunding or common bill splitting is what custom HTLC TLVs would be very nice to experiment with.

alecchendev replied:
Hmm, I'm not sure doing authentication with custom TLVs within a multi-payer single-payee MPP payment is possible in the current implementation of custom TLVs as we drop/fail non-matching TLVs, so if different payment parts had different secrets, they'd just get dropped?

Although, either way, I think for now this recommendation may be a little out of scope for the docs here

ariard replied:
Well you would need to have custom TLVs for `multipayer_secret` where the value is different for each payment parts, and application-level logic to account the TLV. Though I’ve not looked if it’s feasible and working well with current event model.

The hard thing is it’s hard to see if authentication or privacy is screwed or not before you have both a specification of the protocol and implementation, I think the sentence “You MAY authenticate the even TLVs additionally of `payment_secret` to match your protocol semantic” might be reasonable.

alecchendev replied:
Sorry was late to respond to this, is it helpful to add this even though a user couldn't actually take the suggestion? Ultimately any different custom TLV values between different payment parts will be dropped/failed through `RecipientOnionFields::check_merge` in our MPP validation, so the user wouldn't be able to authenticate even TLVs separately from payment secret. If in the future our custom TLV logic were changed to provide the user with individual payment parts' custom TLVs, then this seems fair, but as the code currently stands I'm not sure it makes sense to include?

ariard replied:
> Ultimately any different custom TLV values between different payment parts will be dropped/failed through RecipientOnionFields::check_merge in our MPP validation, so the user wouldn't be able to authenticate even TLVs separately from payment secret. 

Fair point, multi-payer secret would need change in our MPP validation, well kinda restraint what you can do in term of custom HTLC TLVs, though can be extended in the future.

ariard commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -455,7 +459,46 @@ impl RecipientOnionFields {
+		Ok(self)
+	}
+
+	/// Gets the custom TLVs that will be sent or have been received.
```
Maybe the mention “have been received” can be dropped as this is method is about custom HTLC TLVs that are yet to be sent as part of `RecipientOnionField`.

ariard replied:
I would note the structure is called `RecipientOnionFields` with the following comment “Information which is provided, encrypted, to the payment recipient when sending HTLCs”.

ariard reviewed (commented):
See the comment about `InvalidOnionPayload`, otherwise sounds good.

ariard commented on lightning/src/events/mod.rs:
```diff
@@ -356,9 +356,19 @@ pub enum Event {
+	/// If [`Event::PaymentClaimable::onion_fields`] is `Some`, and includes custom TLVs with even type
+	/// numbers, you should use [`ChannelManager::fail_htlc_backwards_with_reason`] with
+	/// [`FailureCode::InvalidOnionPayload`] if you fail to decode these TLVs, or
+	/// [`ChannelManager::claim_funds_with_known_custom_tlvs`] if you successfully decode them.
```
Yes I think “understood and handled the contents” is even better as then when custom protocols are rolled out the future-defined invalid behavior can be restrained from this valid input.

There have been already deanonymization leaks in the past due to onion failure message, see PERM|16.

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -355,15 +359,32 @@ struct ReceiveError {
-	IncorrectOrUnknownPaymentDetails = 0x4000 | 15,
+	IncorrectOrUnknownPaymentDetails,
+	/// We failed to process the payload after the onion was decrypted. You may wish to
+	/// use this when receiving custom HTLC TLVs with even type numbers that you don't recognize.
```
(a) I think deanonymization attack based on onion failure message is in LN security model, see the now-deprecated `incorrect_payment_amount` and `final_expiry_too_soon and b) You introduce a probing vector between versions of LDK, the one giving you back an unknown TLV record and the one giving you a valid result for the same custom TLV record (independently of this being a success or error).

In a world with blinded paths, a payer might never be aware of the node_pubkey of the payee, and such malicious onion payload is a valid deanonymization vector. I agree with you in a world pre-blinding paths, we would care less about fingerprinting.

ariard replied:
> That's a different attack - that's revealing if an HTLC is yours vs if its not.

See the other point if you introduce an overlay protocol like cancellable payment, what content you return as an onion message might leak internal state machine, and therefore if it’s revealing if a HTLC is yours.

On the second attack, software user agent and version are considered in browser deanomymization literature (cf. https://leakuidatorplusteam.github.io/preprint.pdf) and it’s a strong technique to break the one whole anonymous set in segment. As a reminder as a successful de-anonymization attack is just multiple cumulating multiple leaks and correlating them until you can probe the result.

alecchendev replied:
> (https://github.com/lightningdevkit/rust-lightning/pull/2308#discussion_r1278165858) I agree with you in a world pre-blinding paths, we would care less about fingerprinting.

> (https://github.com/lightningdevkit/rust-lightning/pull/2308#issuecomment-1659476435) yes, though i’m not sure no returning error is workable on the long-term due to onion bandwidth cost and payment reliability (at least end-to-end), though here more conversation at the spec-level.

Since we return no error for blinded paths (which may be up to further discussion upstream), and pre-blinded paths fingerprinting is outside our security model, is there any action here?

ariard replied:
> Since we return no error for blinded paths (which may be up to further discussion upstream), and pre-blinded paths fingerprinting is outside our security model, is there any action here?

Looking on the code, it sounds we always return a silent `invalid_onion_payload` as suggested i.e `InvalidOnionPayload(None)` here: https://github.com/lightningdevkit/rust-lightning/pull/2308/files#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR4812

So I think it’s good in case of future errors return even for blinded paths in the future we’re good, no need to change the code.

Looking on the documentation of `FailureCode::InvalidOnionPayload`, “if available, the tuple data should include the type number and byte offset in the decrypted byte stream where the failure occurred”. This could be change to say we never return the custom TLV to avoid leaking user agent / version in case of future onion changes.

alecchendev replied:
Just changed to say "may" instead of "should". I'm going to hold off on mentioning leaking user agent/version unless another reviewer chimes in as it still seems to be outside our security model from what I understand

ariard replied:
Well “may” is reasonable and PR should be numbered in release notes, so user can always go back to this conversation X years from now, assuming GH as a platform is still available.

ariard commented:
See comment here: https://github.com/lightningdevkit/rust-lightning/pull/2308#discussion_r1278165858 I would have agree with Matt in a pre-blinded paths world, though now blinding is a reality I think payer-payee fingerprinting should be in our security model.

alecchendev commented:
Fixed and squashed immediately since they were just small nits

ariard commented:
>  this is why blinded paths have no error returned at all.

yes, though i’m not sure no returning error is workable on the long-term due to onion bandwidth cost and payment reliability (at least end-to-end), though here more conversation at the spec-level.

answered on the two comments:
- i still wish we do our best to mask user agent and version from deanonymization attacks like done for browser / tor services
- i still think “authentication” should be a strong suggestion, seen example of multiple payer not compatible with current `payment_secret`

though overall i won’t insist and if / when we have users privacy screwed up on those vectors, i’ll just pointed it back to you guys :)

ariard commented:
> i still wish we do our best to mask user agent and version from deanonymization attacks like done for browser / tor services

here the code is good in fact after looking (though doc confusing), just good if we _don’t_ change it in the future to match BOLT4 bad suggestion imho

alecchendev commented:
Squashed immediately for minor doc change:
```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 968b0765..daa0b34f 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -371,7 +371,7 @@ pub enum FailureCode {
        /// We failed to process the payload after the onion was decrypted. You may wish to
        /// use this when receiving custom HTLC TLVs with even type numbers that you don't recognize.
        ///
-       /// If available, the tuple data should include the type number and byte offset in the
+       /// If available, the tuple data may include the type number and byte offset in the
        /// decrypted byte stream where the failure occurred.
        InvalidOnionPayload(Option<(u64, u16)>),
 }
```

ariard commented:
Happy after following diff applied, `payment_secret` does not work for N-payer / N-payee flows, though it sounds reasonable for one payer / one payer flow, as far as I know about things like cancellable payments

```
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index daa0b34f..db1108b4 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -4769,6 +4769,8 @@ where
        /// You MUST check you've understood all even TLVs before using this to
        /// claim, otherwise you may unintentionally agree to some protocol you do not understand.
        ///
+       /// You MAY authenticate the even TLVs additionally of `payment_secret` if the protocol
+       /// flow is more sophisticated than one payer / one payee.
        /// [`claim_funds`]: Self::claim_funds
        pub fn claim_funds_with_known_custom_tlvs(&self, payment_preimage: PaymentPreimage) {
                self.claim_payment_internal(payment_preimage, true);
```

ariard reviewed (approved):
Code Review ACK 1db481f94d

Nevermind the above comment effectively multi-payer scheme would need additional code changes in our MPP validation and as such wider custom HTLC support (though they’re the ones more exposed in term of potential deanonymization attacks)

alecchendev commented:
Rebased

alecchendev commented:
Thanks for all the review everybody, will get a follow up up soon :)

alecchendev commented on lightning/src/ln/msgs.rs:
```diff
@@ -1979,15 +1981,23 @@ impl Writeable for OutboundOnionPayload {
+				let preimage = if let Some(ref preimage) = keysend_preimage {
+					Some((5482373484, preimage.encode()))
+				} else { None };
+				let mut custom_tlvs: Vec<&(u64, Vec<u8>)> = custom_tlvs.iter().chain(preimage.iter()).collect();
```
Though AFAIK it wouldn't be possible to sort the tlvs as an iterator?
